File: python/ParentImputerByProgeny.py
from __future__ import annotations

# coding: utf-8
# ParentImputerByProgeny.py
# 後代がimputedで片親がknownな家系の補完
# 後代が複数だと難しいので一つでimputedとする

from math import log
from typing import List, Tuple

from VCFHMM import *
from Genotype import Genotype


#################### ParentImputerByProgeny ####################

MIN_PROB = -1e300

class ParentImputerByProgeny(VCFHMM[VCFRecord]):
	DP = List[Tuple[float, int]]	# (log of probability, prev h)
	
	def __init__(self, records: list[VCFRecord], ref_haps: list[list[int]],
							is_mat_known: bool, map_: Map, w: float) -> None:
		VCFHMM.__init__(self, records, map_)
		self.records: list[VCFRecord] = records
		self.ref_haps = ref_haps
		self.is_mat_known = is_mat_known
		self.prev_h_table = self.collect_possible_previous_hidden_states()
		# crossover values
		# 後代
		self.Cc = [ Map.Kosambi(self.dist(r1, r2))
							for r1, r2 in zip(records, records[1:]) ]
		K = 5						# 親の遷移確率の子に対する倍率
		# 親は直接の後代でないので、乗り換え確率を高くする
		self.Cp = [ Map.Kosambi(self.dist(r1, r2) * K)
							for r1, r2 in zip(self.records, self.records[1:]) ]
	
	def NH(self) -> int:
		return len(self.ref_haps)
	
	# ハプロタイプを決めたときのGenotype
	def gt_by_haplotypes(self, hc1: int, hc2: int,
								mat_gt: int, pat_gt: int) -> int:
		return ((mat_gt >> hc1) & 1) | (((pat_gt >> hc2) & 1) << 1)
	
	def compute_parent_gt(self, i: int, h: int) -> int:
		hw = h & 1			# 後代のどちらに親から渡ったか
		hc = (h >> 1) & 1	# 親のどちらから後代に渡ったか
		hp = h >> 2			# 後代に渡っていない方
		prog_allele = int(self.records[i].v[10][hw*2])	# 親から渡ったアレル
		if hc == 0:
			return prog_allele | (self.ref_haps[hp][i] << 1)
		else:
			return self.ref_haps[hp][i] | (prog_allele << 1)
	
	# i: record index
	def emission_probability(self, i: int, h: int) -> float:
		parent_phased_gt = self.compute_parent_gt(i, h)
		parent_gt = Genotype.gt_to_int(self.records[i].v[9])
		return self.E[parent_phased_gt][parent_gt]	# parent emission
	
	def transition_probability(self, i: int, prev_h: int, h: int) -> float:
		cc = self.Cc[i-1]	# 後代の遷移確率
		cp = self.Cp[i-1]	# 親の遷移確率
		hc = (h >> 1) & 1
		hp = h >> 2
		prev_hc = (prev_h >> 1) & 1
		prev_hp = prev_h >> 2
		
		# 遷移確率 0なら(1-c)、1ならcを掛ける
		return (log(cc if prev_hc != hc else 1.0 - cc) +
				log(cp if prev_hp != hp else 1.0 - cp))
	
	def initialize_dp(self) -> list[DP]:
		M = len(self.records)	# マーカー数
		L = self.NH() << 2
		dp = [ [ (MIN_PROB, 0) ] * L for _ in range(M) ]
		for h in range(L):		# hidden state
			E_all = self.emission_probability(0, h)
			dp[0][h] = (E_all, h)
		return dp
	
	# hidden stateに対して、可能な前のhidden stateを集めておく
	def collect_possible_previous_hidden_states(self) -> list[list[int]]:
		L = self.NH() << 2
		prev_h_table: list[list[int]] = [ [] for _ in range(L) ]
		for h in range(L):		# hidden state
			# 後代のどちらに親から渡ってくるか
			# これは不変
			hw = h & 1
			hc = (h >> 1) & 1	# 親のどちらが渡ってくるか
			hp = h >> 2			# 親の後代に渡ってこない方
			# 複数が同時に乗り換えることはないとする
			for prev_h in range(hw, L, 2):
				prev_hc = (prev_h >> 1) & 1
				prev_hp = prev_h >> 2
				if prev_hp == hp or prev_hc == hc:
					prev_h_table[h].append(prev_h)
		
		return prev_h_table
	
	def update_dp(self, i: int, dp: list[DP]) -> None:
		L = self.NH() << 2
		for h in range(L):		# hidden state
			E_all = self.emission_probability(i, h)
			for prev_h in self.prev_h_table[h]:
				Tp = self.transition_probability(i, prev_h, h)
				prob = dp[i-1][prev_h][0] + (Tp + E_all)
				dp[i][h] = max(dp[i][h], (prob, prev_h))
	
	def update_genotypes(self, hs: list[int]) -> None:
		M = len(self.records)
		# 後代のどちらに渡ったかでGenotypeが入れ替わる可能性がある
		is_swapped = self.is_mat_known ^ (hs[0] & 1) == 0
		for i in range(M):
			record = self.records[i]
			parent_gt = self.compute_parent_gt(i, hs[i])
			record.v[9] = Genotype.int_to_phased_gt(parent_gt)
			if is_swapped:
				gt = record.v[10]
				record.v[10] = gt[2] + '|' + gt[0] + gt[3:]
	
	def impute(self) -> None:
		# DP
		M = len(self.records)
		dp = self.initialize_dp()
		for i in range(1, M):
			self.update_dp(i, dp)
		
		hs = self.trace_back(dp)
		self.update_genotypes(hs)

File: src/ParentImputerByProgeny.cpp
#include <cmath>
#include "../include/ParentImputerByProgeny.h"
#include "../include/common.h"

using namespace std;

ParentImputerByProgeny::ParentImputerByProgeny(const vector<VCFRecord *>& rs,
							 const vector<vector<int>>& ref_hs,
							 bool is_mat_known_, const Map& map_, double w) :
					VCFHMM(rs, map_, w), ref_records(rs), ref_haps(ref_hs),
					prev_h_table(collect_possible_previous_hidden_states()),
					is_mat_known(is_mat_known_),
					Cc(calc_Cc(rs)), Cp(calc_Cp(rs)) { }

vector<double> ParentImputerByProgeny::calc_Cc(
							const vector<VCFRecord *>& rs) const {
	const size_t	M = rs.size();
	vector<double>	Cc(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cc[i] = Map::Kosambi(dist(rs[i], rs[i+1]));
	}
	return Cc;
}

vector<double> ParentImputerByProgeny::calc_Cp(
							const vector<VCFRecord *>& rs) const {
	const double	K = 5.0;
	const size_t	M = rs.size();
	vector<double>	Cp(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cp[i] = Map::Kosambi(dist(rs[i], rs[i+1]) * K);
	}
	return Cp;
}

int ParentImputerByProgeny::compute_parent_gt(size_t i, int h) const {
	const int	hw = h & 1;			// Which of the two haplotypes was passed
									// from the parent to the progeny
	const int	hc = (h >> 1) & 1;	// Which parent passed it to the progeny
	const int	hp = h >> 2;		// Those who have not passed to the progeny
	const string&	prog_gt = records[i]->get_gt(1);
	const int	prog_allele = static_cast<int>(prog_gt.c_str()[hw*2] - '0');
	if(hc == 0)
		return prog_allele | (ref_haps[hp][i] << 1);
	else
		return ref_haps[hp][i] | (prog_allele << 1);
}

double ParentImputerByProgeny::emission_probability(size_t i, int h) const {
	const int	parent_phased_gt = compute_parent_gt(i, h);
	const int	parent_gt = Genotype::gt_to_int(records[i]->get_GT(0));
	return E[parent_phased_gt][parent_gt];
}

double ParentImputerByProgeny::transition_probability(size_t i,
													int prev_h, int h) const {
	const double	cc = Cc[i-1];
	const double	cp = Cp[i-1];
	const int	hc = (h >> 1) & 1;
	const int	hp = h >> 2;
	const int	prev_hc = (prev_h >> 1) & 1;
	const int	prev_hp = prev_h >> 2;
	return log(prev_hc != hc ? cc : 1.0 - cc) +
		   log(prev_hp != hp ? cp : 1.0 - cp);
}

vector<ParentImputerByProgeny::DP>
					ParentImputerByProgeny::initialize_dp() const {
	const size_t	L = NH() << 2;
	vector<DP>	dp(M(), DP(L, pair<double, int>(MIN_PROB, 0)));
	for(int h = 0; h < (int)L; ++h) {
		const double	E_all = emission_probability(0, h);
		dp[0][h] = make_pair(E_all, h);
	}
	return dp;
}

// Collect possible previous hidden states for the hidden state.
vector<vector<int>>
ParentImputerByProgeny::collect_possible_previous_hidden_states() const {
	const size_t	L = NH() << 2;
	vector<vector<int>>	prev_h_table(L, vector<int>());
	for(int h = 0; h < (int)L; ++h) {	// hidden state
		const int	hw = h & 1;		// invariant
		const int	hc = (h >> 1) & 1;
		const int	hp = h >> 2;
		// Neither side will crossover
		for(int prev_h = hw; prev_h < (int)L; prev_h += 2) {
			const int	prev_hc = (prev_h >> 1) & 1;
			const int	prev_hp = prev_h >> 2;
			if(prev_hp == hp || prev_hc == hc)
				prev_h_table[h].push_back(prev_h);
		}
	}
	
	return prev_h_table;
}

void ParentImputerByProgeny::update_dp(size_t i, vector<DP>& dp) const {
	const size_t	L = NH() << 2;
	for(int h = 0; h < (int)L; ++h) {
		const double	E_all = emission_probability(i, h);
		for(auto p = prev_h_table[h].begin(); p != prev_h_table[h].end(); ++p) {
			const int	prev_h = *p;
			const double	Tp = transition_probability(i, prev_h, h);
			const double	prob = dp[i-1][prev_h].first + (Tp + E_all);
			dp[i][h] = max(dp[i][h], make_pair(prob, prev_h));
		}
	}
}

void ParentImputerByProgeny::update_genotypes(const vector<int>& hs) {
	// Genotypes may be swapped depending on
	// which of the progeny passed the parental haplotype
	const bool	is_swapped = is_mat_known ^ ((hs[0] & 1) == 0);
	for(size_t i = 0; i < this->M(); ++i) {
		VCFRecord	*record = ref_records[i];
		const int	parent_gt = compute_parent_gt(i, hs[i]);
		record->set_GT(0, Genotype::int_to_phased_gt(parent_gt));
		if(is_swapped) {
			string&	gt = record->get_mut_gt(0);
			std::swap(gt[0], gt[2]);
		}
	}
}

void ParentImputerByProgeny::impute() {
	// DP
	vector<DP>	dp = initialize_dp();
	for(int i = 1; i < (int)this->M(); ++i) {
		update_dp(i, dp);
	}
	
	const vector<int>	hs = trace_back(dp);
	update_genotypes(hs);
}
